### PR TITLE
redis: auth support

### PR DESCRIPTION
**Pull request by msukalski**
Signed-off-by: Mitch Sukalski <mitch.sukalski@workday.com>

For an explanation of how to fill out the fields, please see the relevant section 
in [PULL_REQUESTS.md](./PULL_REQUESTS.md)

Description:

Authentication support (a la Redis AUTH commands) has been added for both downstream clients and upstream Redis servers on a per-cluster basis. 

A new redis_proxy filter option, downstream_auth_password, can be configured, and no redis commands, including the locally processed PING, will be executed unless a successful authentication command has been issued by the downstream client. Various errors are returned when the authentication command is invalid, or has been issued when the password has not been set; these errors are the same as the errors returned by a Redis server in the same circumstances.

Separately from downstream authentication, the envoy admin can configure an upstream server authentication password on a per-cluster basis. Upstream server connections are transparently authenticated upon connection, if the password has been configured.

Risk Level: low

Testing: unit tests, integration tests, and manual testing against a Redis cluster

Docs Changes:

Changes to version_history.rst under version "1.11.0 (Pending)", and the addition of AUTH support and explanatory text to redis.rst under "Features of Envoy Redis", Supported commands, and Failure modes

Release Notes:

* redis: added auth support :ref:`downstream_auth_password <envoy_api_field_config.filter.network.redis_proxy.v2.RedisProxy.downstream_auth_password>` for downstream client authentication, and :ref:`auth_password <envoy_api_field_config.filter.network.redis_proxy.v2.RedisProtocolOptions.auth_password>` to configure authentication passwords for upstream server clusters.

[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by msukalski** [comment]
@mattklein123 @HenryYYang @FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg @maximebedard I decided to create a PR so folks can review and provide feedback, as they desire.

File: source/common/config/filter_json.cc [hunk 1]
```diff
@@ -238,6 +238,10 @@ void FilterJson::translateRedisProxy(
 
   auto* conn_pool = proto_config.mutable_settings();
   JSON_UTIL_SET_DURATION(*json_conn_pool, *conn_pool, op_timeout);
+
+  if (json_config.hasObject("downstream_auth_password")) {
```
Comment (maximebedard) [inline source/common/config/filter_json.cc]: I think you can remove all the JSON parsing logic as [it's deprecated in v2](https://github.com/envoyproxy/envoy/pull/5623#discussion_r248459784).

File: source/extensions/filters/network/redis_proxy/config.cc [hunk 2]
```diff
@@ -49,10 +49,21 @@ Network::FilterFactoryCb RedisProxyFilterConfigFactory::createFilterFactoryFromP
...
+      if (options) {
+        auth_password = options->auth_password();
+      }
+    }
+
     upstreams.emplace(cluster, std::make_shared<ConnPool::InstanceImpl>(
                                    cluster, context.clusterManager(),
                                    Common::Redis::Client::ClientFactoryImpl::instance_,
-                                   context.threadLocal(), proto_config.settings()));
+                                   context.threadLocal(), proto_config.settings(), auth_password));
```
Comment (maximebedard) [inline source/extensions/filters/network/redis_proxy/config.cc]: Curious, but could fetching the password for the upstream cluster be done in the constructor since the connection pool instance has access to everything it needs?

File: source/common/config/filter_json.cc (hunk 1 above)
Comment (msukalski) [reply source/common/config/filter_json.cc]: hmmm, I'll check into that. I think I added it for the config unit test...

File: source/extensions/filters/network/redis_proxy/config.cc (hunk 2 above)
Comment (msukalski) [reply source/extensions/filters/network/redis_proxy/config.cc]: yes, but I ended up with a bazel circular dependency, and decided that it would be best to keep everything in config.cc. I'll look into it more, and see if I can tease apart the problem.

File: source/common/config/filter_json.cc (hunk 1 above)
Comment (msukalski) [reply source/common/config/filter_json.cc]: It's an integral part of proxy_filter_test.cc at the moment. The test calls translateRedisProxy() directly to create a RedisProxy proto config (in parseProtoFromJson()), so it's not using the deprecated FactoryBase::createFilterFactory() just some of its machinery...

File: source/extensions/filters/network/redis_proxy/config.cc (hunk 2 above)
Comment (msukalski) [reply source/extensions/filters/network/redis_proxy/config.cc]: I've teased it apart, and just pushed new changes. The authentication password is now set in the ThreadLocalPool constructor.

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 3]
```diff
@@ -141,4 +141,23 @@ message RedisProxy {
...
+
+  // Authenticate Redis client connections locally by forcing downstream clients to issue a 'Redis
+  // AUTH command <https://redis.io/commands/auth>`_ with this password before enabling any other
+  // command. If an AUTH command's password matches this password, an "OK" response will be returned
+  // to the client. If the AUTH command password does not match this password, then an "ERR invalid
+  // password" error will be returned. If any other command is received before AUTH when this
+  // password is set, then a "NOAUTH Authentication required." error response will be sent to the
+  // client. If an AUTH command is received when the password is not set, then an "ERR Client sent
+  // AUTH, but no password is set" error will be returned.
+  string downstream_auth_password = 6;
```
[the remark this answers is not in the document]
Comment (msukalski) [reply api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: I'll look into the core.DataSource -- as far as your opening questions are concerned, while the auth command is not very rigorous by itself, it is part of a defense in depth strategy for deploying Redis (TLS encryption is another part of that strategy). Not supporting auth will simply prevent folks from deploying Envoy ... at least at Workday (and I imagine many other firms with sensitive information). In this implementation passwords can be updated incrementally, but we would need to use the hot restart capability to rotate passwords. It would be nice to have these passwords as part of the Secret Discovery Service, but that's a whole another level of infrastructure change. Perhaps the core.DataSource is a good way to straddle the static and fully dynamic worlds?

[the remark this answers is not in the document]
Comment (msukalski) [reply api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: I've changed the passwords to be core.DataSource's instead of strings.

**Comment by msukalski** [comment]
/test/extensions/filters/network/kafka:serialization_test failed, but passes locally with standard debug flags...

**Comment by msukalski** [comment]
@HenryYYang and @FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg -- I've added auth support to the discovery session for the new RedisCluster. There's also an integration test for this support as well.

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 4]
```diff
@@ -7,17 +7,39 @@
...
 namespace Extensions {
 namespace NetworkFilters {
 namespace RedisProxy {
 namespace ConnPool {
 
+namespace {
+
+Common::Redis::Client::DoNothingPoolCallbacks null_pool_callbacks;
+
+Common::Redis::RespValue makeAuthCommand(const std::string& password) {
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: Any difference between this and `Extensions::NetworkFilters::Common::Redis::RespValue makeAuthCommand(const std::string& password)` in `source/extensions/clusters/redis/redis_cluster.cc`?

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 5]
```diff
@@ -218,6 +249,10 @@ InstanceImpl::ThreadLocalPool::makeRequestToHost(const std::string& host_address
     client->redis_client_ =
         parent_.client_factory_.create(it->second, dispatcher_, parent_.config_);
     client->redis_client_->addConnectionCallbacks(*client);
+    if (!auth_password_.empty()) {
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: Nit: Is it worth breaking the whole block `ThreadLocalActiveClientPtr...` into a function `authIfNoClient`?

File: source/extensions/filters/network/redis_proxy/proxy_filter.cc [hunk 6]
```diff
@@ -16,10 +17,14 @@ namespace RedisProxy {
...
     const envoy::config::filter::network::redis_proxy::v2::RedisProxy& config, Stats::Scope& scope,
-    const Network::DrainDecision& drain_decision, Runtime::Loader& runtime)
+    const Network::DrainDecision& drain_decision, Runtime::Loader& runtime, Api::Api& api)
     : drain_decision_(drain_decision), runtime_(runtime),
       stat_prefix_(fmt::format("redis.{}.", config.stat_prefix())),
-      stats_(generateStats(stat_prefix_, scope)) {}
+      stats_(generateStats(stat_prefix_, scope)),
+      downstream_auth_password_(
+          Config::DataSource::read(config.downstream_auth_password(), true, api)) {
+  auth_required_ = !downstream_auth_password_.empty();
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline source/extensions/filters/network/redis_proxy/proxy_filter.cc]: Nit: This isn't used much- easier to just check for `connection_allowed_ = config-downstream_auth_password.empty()`?

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc (hunk 4 above)
Comment (msukalski) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: no ... I'll create a single instance in Extensions::NetworkFilters::Common::Redis::Utility namespace (and add a new utility.cc and utility.h to extensions/filters/network/common/redis)...

File: source/extensions/filters/network/redis_proxy/proxy_filter.cc (hunk 6 above)
Comment (msukalski) [reply source/extensions/filters/network/redis_proxy/proxy_filter.cc]: yeah, I missed that -- thanks. It was originally going to be more complicated...sigh...

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc (hunk 5 above)
Comment (msukalski) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: I'll move all of this logic into a threadLocalActiveClient() method for InstanceImpl::ThreadLocalPool ...

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
LGTM. As discussed key rotation is not supported by Redis anyway, and Redis 6 "changes things", so this works fine.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
👍

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
@mattklein123 any thoughts?

**Comment by msukalski** [comment]
@mattklein123 Can I get someone to approve this PR so it can be merged? The current PR has addressed all review feedback from folks...

**Comment by mattklein123** [comment]
Will take a look.

File: source/common/json/config_schemas.cc [hunk 7]
```diff
@@ -441,7 +441,7 @@ const std::string Json::Schema::REDIS_PROXY_NETWORK_FILTER_SCHEMA(R"EOF(
       "conn_pool" : {"type" : "object"}
     },
     "required": ["cluster_name", "stat_prefix", "conn_pool"],
-    "additionalProperties": false
+    "additionalProperties": true
```
Comment (mattklein123) [inline source/common/json/config_schemas.cc]: We shouldn't be making any v1 changes anymore so we should be able to revert this change.

File: source/extensions/filters/network/redis_proxy/config.h [hunk 8]
```diff
@@ -13,11 +15,31 @@ namespace Extensions {
...
+  std::string auth_password(Api::Api& api) const {
+    return Config::DataSource::read(auth_password_, true, api);
+  }
+
+  const envoy::api::v2::core::DataSource& auth_password_datasource() const {
+    return auth_password_;
+  }
+
+private:
+  envoy::api::v2::core::DataSource auth_password_;
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/config.h]: nit: const here and in similar places if possible.

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.h [hunk 9]
```diff
@@ -93,6 +99,7 @@ class InstanceImpl : public Instance {
     std::unordered_map<Upstream::HostConstSharedPtr, ThreadLocalActiveClientPtr> client_map_;
     Envoy::Common::CallbackHandle* host_set_member_update_cb_handle_{};
     std::unordered_map<std::string, Upstream::HostConstSharedPtr> host_address_map_;
+    std::string auth_password_;
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.h]: const

File: source/common/json/config_schemas.cc (hunk 7 above)
Comment (msukalski) [reply source/common/json/config_schemas.cc]: ok .. I'll change the proxy_filter_test.cc code to add config that is not v1 to avoid the schema validation exception.

File: source/extensions/filters/network/redis_proxy/config.h (hunk 8 above)
Comment (msukalski) [reply source/extensions/filters/network/redis_proxy/config.h]: ok

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.h (hunk 9 above)
Comment (msukalski) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.h]: ok ... but this will require explict const_cast usage in the constructor to set it initially

File: test/extensions/filters/network/redis_proxy/proxy_filter_test.cc [hunk 10]
```diff
@@ -38,6 +38,12 @@ parseProtoFromJson(const std::string& json_string) {
   envoy::config::filter::network::redis_proxy::v2::RedisProxy config;
   auto json_object_ptr = Json::Factory::loadFromString(json_string);
   Config::FilterJson::translateRedisProxy(*json_object_ptr, config);
+  return config;
+}
+
+void addV2ProtoFromJson(const std::string& json_string,
```
Comment (mattklein123) [inline test/extensions/filters/network/redis_proxy/proxy_filter_test.cc]: We are killing v1. Can you just switch all the tests over to v2 config and delete all v1 code from this test? cc @derekargueta

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 11]
```diff
@@ -56,7 +56,7 @@ InstanceImpl::ThreadLocalPool::ThreadLocalPool(InstanceImpl& parent, Event::Disp
     auto options = cluster->info()->extensionProtocolOptionsTyped<ProtocolOptionsConfigImpl>(
         NetworkFilterNames::get().RedisProxy);
     if (options) {
-      auth_password_ = options->auth_password(parent_.api_);
+      const_cast<std::string&>(auth_password_) = options->auth_password(parent_.api_);
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: Sorry please revert this part of the change. If you can't set it inside an initializer list just leave it non-const.

File: test/extensions/filters/network/redis_proxy/proxy_filter_test.cc (hunk 10 above)
Comment (msukalski) [reply test/extensions/filters/network/redis_proxy/proxy_filter_test.cc]: I'm pretty sure that the only V1 code is translateRedisProxy(), so I'll remove that. I'm going to directly call some of the JSON utility macros, but they all appear to be v2 macros.

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc (hunk 11 above)
Comment (msukalski) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: ok

**Comment by msukalski** [comment]
It looks like the coverage failure is due to "Code coverage 97.4 is lower than limit of 97.5". None of my code modifications do not have code coverage, so it looks like I'll have to artificially cover non-functional interface methods for RedisActiveHealthCheckSession, RedisDiscoveryClient, and RedisDiscoverySession to get line coverage back up to the limit.

**Comment by mattklein123** [comment]
You can add coverage of more redis code, or look https://s3.amazonaws.com/lyft-envoy/coverage/report-master/coverage.html for other coverage to add. There is quite a bit of legit coverage missing elsewhere.

/wait

**Comment by HenryYYang** [comment]
+1

**Comment by msukalski** [comment]
ci/circleci:docs check failed due to "checking consistency... /source/generated/rst/configuration/best_practices/edge.rst: WARNING: document isn't included in any toctree" -- I didn't add this file, so I don't know how it got merged into master passing this check. Sigh.

**Comment by mattklein123** [comment]
@msukalski it was an accidental force merge. It will get fixed in the morning. cc @htuch 

/wait
